### ISSUE TITLE
5% regression in some Inference benchmarks due to #51149 (Set temporary env whenever Julia is started with `--project=@temp`)

### ISSUE THREAD

**Issue by Zentrik** [JuliaLang/julia#54850]
I'm confused as to how 583981fb43c7480f45335fe7ebc73aaaa74c27d0 has caused this, but it caused a 6% regression in min wall time, allocations and memory for the `[["inference", "allinference", "REPL.REPLCompletions.completions"]]` benchmark and a 3% regression for the `[["inference", "allinference", "Base.init_stdio(::Ptr{Cvoid})"]]` becnhmark.

I would presume this regression should be fixable given how little the pr changed but I haven't been able to figure out what caused the regression in the first place.

Nanosoldier report: https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_date/2024-06/10/report.md

**Comment by oscardssmith** [comment]
Are you sure this isn't noise?

**Comment by Zentrik** [comment]
Yes, I can reproduce it reliably and it reproduced on the only other nanosoldier run, see https://tealquaternion.camdvr.org/compare.html?start=2024-06-04&end=2024-06-13&stat=min-wall-time&name=allinferen.

**Comment by IanButterworth** [comment]
Sorry if I'm being slow, but what from the reports indicates that commit? There's a huge number of commits between that report and the previous?

**Comment by Zentrik** [comment]
Sorry, I bisected the regression to that commit.

**Comment by KristofferC** [comment]
> and a 3% regression for the [["inference", "allinference", "Base.init_stdio(::Ptr{Cvoid})"]] benchmark.

This seems like dup of the reason in https://github.com/JuliaLang/julia/issues/53695, https://github.com/JuliaLang/julia/issues/53459#issuecomment-1989647216

**Comment by Zentrik** [comment]
Yes, for both benchmarks it's because it now needs to infer `mktempdir` in `load_path_expand`.
